socorro/cron/jobs/first_signatures.py

```diff
@@ -0,0 +1,204 @@
+@as_backfill_cron_app
+class FirstSignaturesCronApp(BaseCronApp):
+    app_name = 'first-signatures'
+    app_description = 'Using SuperSearch to log when signatures first appeared'
+    app_version = '0.1'
+
+    required_config = Namespace()
+    required_config.add_option(
+        'window_seconds',
+        default=60,  # seconds
+        doc='Number of seconds to search for signature facets back'
+    )
+    required_config.add_option(
+        'sensitive_to_errors',
+        # default=True,
+        default=False,
+        doc=(
+            'If there are, for example, missing indexes in ES you can get errors '
+            "in supersearch doing queries across dates we don't have indices for. "
+            'If this is the case, halt this crontabber app.'
+        )
+    )
+    required_config.add_option(
+        'historic_days',
+        # The number 26 comes from the default in index_cleaner which is used
+        # as a retention policy for cleaning elasticsearch.
+        default=26 * 7,  # days
+        doc=(
+            'When looking for the oldest crash possible (by a specific signature) '
+            'go back this many days'
+        )
+    )
+    required_config.add_option(
+        'elasticsearch_class',
+        default=ElasticsearchConfig,
+    )
+
+    def run(self, date):
+        # First get a list of all unique signatures, in the given timespan
+        all_fields = SuperSearchFields(config=self.config).get()
+        api = SuperSearch(config=self.config)
+        assert isinstance(date, datetime.datetime), type(date)
+        start_date = date - datetime.timedelta(seconds=self.config.window_seconds)
+        params = {
+            'date': [
+                '>={}'.format(start_date.isoformat()),
+                '<{}'.format(date.isoformat()),
+            ],
+            '_facets': 'signature',
+            '_facets_size': 1000,
+            '_results_number': 0,
+            '_fields': all_fields,
+        }
+        results = api.get(**params)
+        signatures = [
+            x['term'] for x in results['facets']['signature']
+        ]
+        self.config.logger.info(
+            'Found %s unique signatures between %s and %s',
+            len(signatures),
+            start_date,
+            date,
+        )
+        if not signatures:
+            self.config.logger.info(
+                'Exit early because no signature facets'
+            )
+            return
+
+        # Now to figure out which ones we've never seen before
+        signature_first_date_api = SignatureFirstDate(config=self.config)
+        results = signature_first_date_api.get(signatures=signatures)
+        new_signatures = set(signatures) - set(results['hits'])
+        self.config.logger.info('%s new signatures', len(new_signatures))
+
+        # Now, for each of these new signatures we have to find the oldest
+        # crash that has it and write down its date and buildid.
+        historic_date = date - datetime.timedelta(
```

Comment:
You do not need to fetch historical data. You can trust that the table has correct data up until an hour ago, so if something isn't in that table yet, you can be sure that it only appeared in the last hour.

Reply:
I don't get it. I really want the oldest crash possible. I sort by oldest first. But that's just oldest first in the time window. So if I don't use historical data, I'd get the oldest of the 1hour most recent crashes. 
No?

Reply:
Yes, but since that signature was not in the ``signature`` table beforehand, that means its first appearance date has to be in the time window of the currently running job. 

Say you have this in your ``signatures`` table: 

signature | date | build
-|-|-
foo | 2010-01-01 | 20100101
bar | 2016-10-12 | 20161010

Now you are running that job for the hour between ``2017-01-01 00:00:00`` and ``2017-01-01 01:00:00``, and you find a signature that doesn't belong to the table, ``bean``. Because you trust that the data in the ``signatures`` table is correct, you know that that ``bean`` signature never appeared before. Hence its first appearance date, and first build, has to be contained within the date range of the job, between ``2017-01-01 00:00:00`` and ``2017-01-01 01:00:00``. There is no need to look for it in older data, because the ``signatures`` table tells you that it never appeared before.

Reply:
Ah! Makes sense now. I get it. Thanks!